### PR TITLE
Typing the untyped

### PR DESCRIPTION
**Pull request by tranma**
Provides:
- parsing/printing from/to SimplePoints to/from our decoded payload types via prisms.
- folding decoded points for each resource. There are common fold methods these resource-specific folds could re-use, it's also easy to add resource-specific edge cases:
  
  ```
      -- Stop folding once the volume is deleted.
      step   (More b) (Timed _ (PDVolume _ VolumeDelete _ _)) = Term b
      -- Measure the duration the volume spends at each size.
      step a x = stepCompoundEvent window a x
  ```
- infer the prism/fold to use based on the schema (SourceDict)
- convenient client interface

### CODE REVIEW

File: lib/Ceilometer/Types.hs [hunk 1]
```diff
@@ -1 +1,111 @@
...
+
+type Env = (FlavorMap, SourceDict)
+
+lookupMetricName = lookupSource "metric_name"
+lookupEvent      = lookupSource "_event"
+
+type CSimple = SimplePoint
+
+-- | Views the raw @SimplePoint@ as-is.
+cSimple :: Env -> Prism' SimplePoint CSimple
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: prism' id f seems odd to me unless you're actually planning to do something in the backwards direction.

What's going on here is a little misleading. I don't like the idea of a type synonym here, perhaps a newtype? If all you want to do is have a partial composable reviewable thing, you may be looking for a Fold.

``` haskell
cSimple :: Env -> Fold SimplePoint CSimple
csSimple (_, sd) = folding parse
  where ...
```

Then you can review it, because that works on folds.

File: lib/Ceilometer/Types.hs [hunk 2]
```diff
@@ -1 +1,111 @@
...
+import           Data.Word
+
+import           Marquise.Client
+import           Vaultaire.Types
+
+import           Ceilometer.Types.Base
+import           Ceilometer.Types.IP
+import           Ceilometer.Types.Volume
+
+import Control.Lens.Prism
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: Pls to use git-vogue.

File: lib/Ceilometer/Types.hs [hunk 3]
```diff
@@ -1 +1,111 @@
...
+          _                               -> Nothing
+
+data CVolume = CVolume Address TimeStamp VolumeStatus VolumeVerb Endpoint Word64
+
+-- | Views the raw @SimplePoint@ as a Ceilometer point @CVolume@.
+cVolume :: Env -> Prism' SimplePoint CVolume
+cVolume (flavours, sd) = prism' toSP fromSP
+  where toSP = undefined
+        fromSP (SimplePoint a ts v) = case (lookupMetricName sd, isEvent sd) of
+          (Just "volume.size", False) -> Just $ CVolume a ts _ _ _ _
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: This can't be safe? Is this complete?

File: lib/Ceilometer/Types.hs [hunk 4]
```diff
@@ -1 +1,111 @@
...
+
+data CIPEvent = CIPEvent Address TimeStamp IPStatus IPVerb Endpoint
+
+-- | Views the raw @SimplePoint@ as a partially applied `step`
+--   to use in an aggregation of @SimplePoint@.
+--   Assuming the aggreation is a fold.
+step :: Prism' SimplePoint (acc -> acc)
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: Not sure how this would be partial, I'll leave this until you do more.

File: lib/Ceilometer/Types.hs [hunk 5]
```diff
@@ -1 +1,111 @@
...
+-- Images
+data ImageStatus
+data ImageVerb
+
+-- Instances
+--
+newtype Flavor = Flavor String
+  deriving (Show, Eq, Ord)
+
+type FlavorMap = Bimap Flavor Word64
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: I like this Bimap thing.

File: lib/Ceilometer/Types.hs [hunk 6]
```diff
@@ -1 +1,111 @@
...
+  = InstancePoint       Address TimeStamp InstanceStatus                       InstancePayload
+  | IPEventPoint        Address TimeStamp IPStatus       IPVerb       Endpoint
+  | VolumeEventPoint    Address TimeStamp VolumeStatus   VolumeVerb   Endpoint Word64
+  | SnapshotEventPoint  Address TimeStamp SnapshotStatus SnapshotVerb Endpoint Word64
+  | ImageEventPoint     Address TimeStamp ImageStatus    ImageVerb             Word64
+  | StandardPoint       Address TimeStamp                                      Word64
+
+--type VaultIso = Iso SimplePoint
+
+isEvent :: SourceDict -> Bool
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: Partial.

File: lib/Ceilometer/Types.hs [hunk 7]
```diff
@@ -1 +1,111 @@
...
+isEvent sd = case lookupSource "_event" sd of
+    Just "1" -> True
+    Just "0" -> False
+    Nothing  -> False
+
+metricName :: SourceDict -> Maybe Text
```
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: lookupSource "metric_name" seems clearer to me as-is.

File: lib/Ceilometer/Types/Base.hs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+module Ceilometer.Types.Base where
+
+import           Control.Isomorphism.Partial.Iso
+import           Data.Bimap                      (Bimap)
+import qualified Data.Bimap                      as B
+import           Data.Word
+
+-- |Iso generator
+generateIsoFromBimap :: (Ord a, Ord b) => Bimap a b -> Iso a b
```
Comment (christian-marie) [inline lib/Ceilometer/Types/Base.hs]: From bijection to isomorphism, neat!

You probably want to use unsafeMakeIso' and stick error messages in there, otherwise if you stick this in a long chain of isos you will not know which one failed. That may require Show instances to make useful instances. If that's not okay, you can just  name it.

File: lib/Ceilometer/Types/Base.hs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+generateIsoFromBimap bm = unsafeMakeIso (`B.lookup` bm) (`B.lookupR` bm)
+
+data Endpoint = Start
+              | End
+  deriving (Show, Eq, Ord)
+
+endpointMap :: Bimap Endpoint Word64
+endpointMap = foldl (\bm (s, v) -> B.insert s v bm)
```
Comment (christian-marie) [inline lib/Ceilometer/Types/Base.hs]: You can't use the fromLists that ship with Bimap? They should do basically the same thing.

File: lib/Ceilometer/Types.hs (hunk 1 above)
Comment (tranma) [inline lib/Ceilometer/Types.hs]: The idea was to define prisms to view raw `SimplePoint` as decoded points, and also going backwards (for some publisher thing) and compose those prisms in a big function below.

I've scraped this idea and will remove the prism code, instead I think we should use plain old Control.Lens.Iso as decoding these points are not like parsing JSON blobs of unknown length and shape.

File: lib/Ceilometer/Types/IP.hs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+import Ceilometer.Types.Base
+
+data IPStatus = IPNone
+              | IPActive
+              | IPDown
+  deriving (Show, Eq, Ord)
+
+ipStatusMap :: Bimap IPStatus Word64
+ipStatusMap = foldl (\bm (s, v) -> B.insert s v bm)
```
Comment (christian-marie) [inline lib/Ceilometer/Types/IP.hs]: This folds keeps happening over and over, it's fromList.

File: lib/Ceilometer/Types.hs (hunk 6 above)
Comment (tranma) [inline lib/Ceilometer/Types.hs]: what is? the big sum type? I don't think that's good either. We should instead restrict the domain and use iso's (I mean real iso's)

File: lib/Ceilometer/Types/Base.hs (hunk 8 above)
Comment (tranma) [inline lib/Ceilometer/Types/Base.hs]: Bijection? a -> b doesn't seem to be a bijective function to me
@oswynb

**Comment by tranma** [comment]
The `Iso` in this codebase is from `roundtrip`'s "partial" isomorphism, I presume. But I think plain old isomorphism is better here, since we're not parsing unknown blobs.

@christian-marie

**Comment by christian-marie** [comment]
Your lookup can fail, hence it must be a partial iso to typecheck unless you stick fromJusts in there.

**Comment by christian-marie** [comment]
> Goals:
> 
> unify serialisation/deserialisation of SimplePoint <-> (some typed Ceilometer
> point) provide an interface for Borel to use (for aggregation)
> 
> Proposal:
> 
> plain old Control.Lens.Iso to define print/parse for the payload of a
> SimplePoint. Afaict, this is the only part we need to decode, and unlike JSON,
> it doesn't need a full-blown partial parser and pretty printer. It's not a blob
> of unknown length and shape.

This seems fine, if you can convince yourself that you will never need
partiality.

> given a (dictionary, stream of SimplePoint), convert to (stream of some type
> point) for Borel to use. This handling of raw SimplePoint streams everywhere
> has given us so much grief, I wonder why Marquise doesn't let you get some
> TypedPoint instead (it seems like we actually do have a schema to define
> TypedPoint - the dictionary, this should always be the case?)

Maybe you want to define a bunch of Folds to help?

https://hackage.haskell.org/package/foldl

> Unclear:
> 
> where the different aggregation methods for each (some typed Ceilometer Point)
> should live. Currently it's all over vaultaire-query, borel, borel-core.  many
> other things???

I'd have stuck it in marquise.

File: lib/Ceilometer/Types.hs (hunk 6 above)
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: Nope, I mean the function below is partial. Just "1", Just "0", but not Just "butterflies"

File: lib/Ceilometer/Types/Base.hs (hunk 8 above)
Comment (christian-marie) [inline lib/Ceilometer/Types/Base.hs]: Well, the bimap is a sort of bijection, if you restrict  image to be only the things in the map.

"For most algebraic structures, including groups and rings, a homomorphism is an isomorphism if and only if it is bijective."

File: lib/Ceilometer/Types.hs (hunk 6 above)
Comment (tranma) [inline lib/Ceilometer/Types.hs]: I don't like how we had to define a bunch of partial predicates like this on what is essentially the schema of our data either. Instead, it would be nice if we can somehow parse this dictionary into a schema of the types we want.

File: lib/Ceilometer/Types/Base.hs (hunk 8 above)
Comment (tranma) [inline lib/Ceilometer/Types/Base.hs]: Yeah, but there is no guarantee the values in the Bimap are unique.

File: lib/Ceilometer/Types.hs (hunk 6 above)
Comment (christian-marie) [inline lib/Ceilometer/Types.hs]: It is the schema of our data, sure, unfortunately once that's gone over the wire we can't rely on it any more than a binary protocol.

I've thought about doing crazy things like defining an abstraction layer in which you put dirty blobs in and you get some nice tagged-and-verified stuff out the other end. It's non-trivial, though.

Perhaps we could use something along the lines of: https://hackage.haskell.org/package/dependent-sum-0.2.1.0/docs/Data-Dependent-Sum.html

Comment (tranma) [inline lib/Ceilometer/Types.hs]: I don't think it would be too hard to add a middle layer that translate raw data into typed Haskell values. We could even generate these types from schemas with TH. It isn't trivial, but neither is managing raw streams with a typeless schema tagging along.

Comment (tranma) [inline lib/Ceilometer/Types.hs]: Not even any thing as fancy as a dependently typed value. Just a typed Haskell value would have sufficed.

File: lib/Ceilometer/Types/Base.hs (hunk 8 above)
Comment (christian-marie) [inline lib/Ceilometer/Types/Base.hs]: We're assuming the domain is a set, yes? In this case, it is because it's already been stuffed into the bimap, deleting any duplicates.

**Comment by olorin** [comment]
> > where the different aggregation methods for each (some typed Ceilometer Point)
> > should live. Currently it's all over vaultaire-query, borel, borel-core. many
> > other things???
> 
> I'd have stuck it in marquise.

Why marquise for Ceilometer-metering-specific stuff? I thought a Vaultaire client library should read stuff from Vaultaire, no need to unnecessarily embiggen it. Thoughts?

(It seems to me that ceilometer-common might be a good place for that. Or borel-core. Either way, better one place than several.)

**Comment by tranma** [comment]
@fractalcat because Marquise is the entry-point to Vaultaire data, hence everyone just fetches raw untyped data and struggles with it. If Marquise gives you typed data in the first place, there'd be less temptation to flail

**Comment by olorin** [comment]
That's true; I'd always envisaged people using a layer on top of Marquise for that kind of thing, like vaultaire-query. I think there are advantages to maintaining a stupid-simple interface that doesn't try to do anything clever; I'm open to it being something else than the marquise package if there's a good case for it.

**Comment by oswynb** [comment]
I've always thought of Marquise was designed to get data from and get data into the vault. Like @fractalcat I've felt that having a layer on top to deal with domain specific typing/handling would be the norm. Could have things like marquise-pmacct or marquise-ceilometer, etc.

Alternatively/in addition to we could add folding functions to run on the broker nodes (I think this was being discussed before? In addition to having simple aggregation on the ceph nodes)

Something like:
foldSimple :: (MarquiseReaderMonad m conn) => Policy -> Address -> TimeStamp -> TimeStamp -> Origin -> conn -> (acc -> SimplePoint -> acc) -> acc -> Marquise m acc

For the specific use-case here ceilometer-common could provide the folding function

type CAcc = Set (CPoint, Word64)
cJoin :: Env -> (CAcc -> CPoint -> CAcc)
cFlatten :: Env -> CAcc -> CAcc

result <- runMarquise <$> cFlatten env <$> M.foldSimple ... (cJoin env)

TL;DR: OK for adding functionality to Marquise, but aggregation methods for Ceilometer should be in ceilometer-common, not marquise or borel

**Comment by tranma** [comment]
That is the thing, we want Marquise to be about getting "data" to and from the vault too, but we want "data" here to be typed, not just a raw stream and a schema.

What you propose essentially delegates the job of using the schema to interpret the raw stream to other packages, and this is fine, except right now no one does that and everyone handles the raw stream. I think it's a lot better to restrict the ability to access untyped data to users. Marquise can provide raw streams too but this should be secondary to typed streams.

I had roughly the same idea with the generic "step" function that supposedly steps a raw `SimplePoint` by decoding that untyped point using the schema client-side. But this is a very Javascript thing to do, in Haskell you really want to get the things typed asap.

I think it comes down to...philosophical preferences? @christian-marie and @fractalcat what do you think?

**Comment by tranma** [comment]
^ this commit prototypes the prisms/iso's approach. I use this instead of roundtrip-style partial iso's because the payload is very well-defined.
@oswynb @christian-marie

**Comment by christian-marie** [comment]
Philosophically, I think that you do your very best to provide an interface that's both intuitive and easy to use and impossible to type things that don't make sense or will break. So I'm always for wrapping things up.

This is a lot of work, is all, and you don't just do it piece by piece.

**Comment by tranma** [comment]
This PR is ready for review. I think it can merged now with the currently defined rsource types (CPU, instance VCPU/RAM/flavors, Volume) and any additional resource (Image, Snapshot, etc.) can be added the same way and re-using most of the logic already defined here. I'd like to start using this with Borel to test its flexibility.

Please review, @fractalcat and @oswynb?

@christian-marie  might be interested in how we used prisms & QC & hpc?

File: lib/Ceilometer/Client.hs [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+-- | Decode and fold a stream of @SimplePoint@.
+--
+--   If the assertion that these points should be type `a` is wrong: run-time error.
+--
+--   If the assertion is true but the @SourceDict@ doesn't match what we expect
+--   for the fold: `Nothing`.
+--
+--   Otherwise we get a `FoldResult` matching that type.
+--
+--   Example usage: `decodeAndFold (undefined :: proxy PDCPU) env points`
```
Comment (christian-marie) [inline lib/Ceilometer/Client.hs:57]: Data.Proxy has a real proxy, such that we don't need to throw around undefineds. We use that in all of our other type magic code. Happy to revisit this when we have better language support.

File: lib/Ceilometer/Client.hs [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+-- | Decodes the raw stream of SimplePoint using the given prism for the payload.
+decodeWith
+  :: Monad m
+  => Prism' Word64 a                         -- ^ Use this particular prism for decoding
+  -> Pipe SimplePoint (Maybe (Timed a)) m () -- ^ Decoding pipe
+decodeWith p = forever $ do
+  SimplePoint _ (TimeStamp t) v <- await
+  yield $ T.sequence $ Timed t $ v ^? p
+
+ignore :: Monad m => Pipe (Maybe x) x m r
```
Comment (christian-marie) [inline lib/Ceilometer/Client.hs:95]: Documentation for this function would be nice, ignore is probably a misnomer.

This seems like a very specific:

``` haskell
concat :: (Monad m, Foldable f) => Pipe (f a) a m r 
```

Flatten all Foldable elements flowing downstream

File: lib/Ceilometer/Client.hs (hunk 11 above)
Comment (tranma) [inline lib/Ceilometer/Client.hs:57]: unless we have Signatures Sections `foo (:: MyType) x y...`, either this or `Data.Proxy` doesn't really make a difference.

File: lib/Ceilometer/Client.hs (hunk 12 above)
Comment (tranma) [inline lib/Ceilometer/Client.hs:95]: not really. We don't "flatten and yield" any Foldable, we flatten, then decide whether to yield or not based on the result.

File: lib/Ceilometer/Client.hs (hunk 11 above)
Comment (christian-marie) [inline lib/Ceilometer/Client.hs:57]: It does in that we're not passing undefined around and everyone is aware of the idiom.

Specifically when you have a typeclass that takes a proxy, you have something to pattern match on that is explicit also, like:

``` haskell
f :: Monoid x => Proxy x -> x
f Proxy = mempty
```

File: lib/Ceilometer/Client.hs (hunk 12 above)
Comment (christian-marie) [inline lib/Ceilometer/Client.hs:95]: I can't find any difference between your definition and concat other than it is more specific.

It "behaves" the same, too.

``` haskell
λ P.toList $ each [Just 1, Nothing, Just 2] >-> ignore
[1,2]
λ P.toList $ each [Just 1, Nothing, Just 2] >-> P.concat
[1,2]
```

File: lib/Ceilometer/Client.hs (hunk 11 above)
Comment (tranma) [inline lib/Ceilometer/Client.hs:57]: I would like this parameter to act like a real type application, and forbid anyone from pattern matching on it (that's why it's a "proxy a" and not "a").

**Comment by tranma** [comment]
went through review with @christian-marie 
more or less sane, merging so travis can go on the other PRs
